### DISCUSSION TITLE
Scheduler is not working

### DISCUSSION THREAD

**Discussion by vengi57** [apache/airflow#17705] [General]
**Apache Airflow version**:

Version - 2.1.1

**OS**:
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 18.04.2 LTS
Release:	18.04
Codename: bionic

**Apache Airflow Provider versions**:

apache-airflow-providers-ftp==2.0.0
apache-airflow-providers-imap==2.0.0
apache-airflow-providers-postgres==2.0.0
apache-airflow-providers-sqlite==2.0.0

**Deployment**:
 VM 

**What happened**:
Getting this warning message
The scheduler does not appear to be running. The last heartbeat was received 2 days ago. The DAGs list may not update, and new tasks will not be scheduled

So that when I trigger my DAG it was stuck in the in-progress state for a long time and none of the  tasks of  DAGs are running

If I restart the scheduler it was working fine for some hours but again getting the same  above mentioned issue 

**What you expected to happen**:
The scheduler should work properly in order to execute the dags which will take around 3 to 5 hrs to complete

Even though no DAGs are running we getting "The scheduler does not appear to be running. The last heartbeat was received 2 days ago. The DAGs list may not update, and new tasks will not be scheduled" error 

**How to reproduce it**:
1.Install Airflow
2. Change the executor to Local Executor
3. Change sql_alchemy_conn to Postgres DB
4. Run airflow webserver &
5. Run airflow scheduler &

**Anything else we need to know**:
This problem occurs when not dags are running for a while and getting  The scheduler does not appear to be running error

Attached screenshot of the issue
![Screenshot from 2021-08-17 11-26-57](https://user-images.githubusercontent.com/32476102/129671390-93edc9e3-22c2-482d-9357-1492819d501f.png)

If I checked the scheduler status with /heath API getting unhealthy for the scheduler
![Screenshot from 2021-08-12 16-52-57](https://user-images.githubusercontent.com/32476102/129674037-3b69bc94-00c2-4535-8b70-47e22e2cb008.png)

Attached airflow.cfg file for the reference
[airflow.txt](https://github.com/apache/airflow/files/6997420/airflow.txt)

**Comment by ephraimbuddy** [comment]
Can you include the scheduler log covering the period that this happened?

**Comment by potiuk** [comment]
Yep. For me it looks like the scheduler could simply crash and you have not restarted it (or for whatever reason it stopped running because of some deadlock or similar). But we need to see the logs. Also  - what is the way you implemented automated restarting of scheduler if it crashed and monitoring? I think if you do not have automated restart nor monitoring implemented, you should do it as  soon as you can.

**Comment by vengi57** [comment]
For now, no automation is implemented for the webserver and scheduler  when any issue happens
Can you please suggest to me a way to implement the auto-restart of the scheduler if any crash or deadlock occurs
Can I use pm2 for starting of scheduler and webserver will it works?

**Comment by potiuk** [comment]
It's up to you - but our suggested (and community sypported) way of deploying Airflow is through Kubernetes and Official Helm Chart  https://airflow.apache.org/docs/helm-chart/stable/index.html. When it comes to "home-grown" way of doing it when you do not want or can afford learning K8S, I have no suggestions - it's whatever you are used to for anything else you already manage. Airflow has nothing "specific" that would make it difficult to integrate with any system, it's a standard tool as many others and you can adapt it to whatever deployment you have.

**Reply by vengi57** [reply]
After using **PM2** for starting airflow webserver and airflow scheduler, the scheduler is not going for a toss and we are not facing "**The scheduler does not appear to be running. The last heartbeat was received 2 days ago. The DAGs list may not update, and new tasks will not be scheduled**" error
Also, **/health** API giving healthy status for the scheduler

**Reply by potiuk** [reply]
Cool!
